src/amd_ipmi.c

```diff
@@ -0,0 +1,454 @@
+/*
+ * AMD IPMI LED control
+ * Copyright (C) 2019, Advanced Micro Devices, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <libgen.h>
+#include <inttypes.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <sys/sysinfo.h>
+#include <sys/file.h>
+
+#if _HAVE_DMALLOC_H
+#include <dmalloc.h>
+#endif
+
+#include "config.h"
+#include "ibpi.h"
+#include "list.h"
+#include "utils.h"
+#include "amd.h"
+#include "ipmi.h"
+
+static uint8_t amd_ibpi_ipmi_register[] = {
+	[IBPI_PATTERN_PFA] = 0x41,
+	[IBPI_PATTERN_LOCATE] = 0x42,
+	[IBPI_PATTERN_FAILED_DRIVE] = 0x44,
+	[IBPI_PATTERN_FAILED_ARRAY] = 0x45,
+	[IBPI_PATTERN_REBUILD] = 0x46,
+	[IBPI_PATTERN_HOTSPARE] = 0x47,
+};
+
+/* The path we are given should be similar to
+ * /sys/devices/pci0000:e0/0000:e0:03.3/0000:e3:00.0
+ *                                      ^^^^^^^^^^
+ * We need to retrieve the address from the path (indicated above)
+ * then use it to find the corresponding address for a slot in
+ * /sys/bus/pci_slots to determine the icorrect port for the NVMe device.
+ */
+static int _get_ipmi_nvme_port(char *path)
+{
+	int rc;
+	char *p, *f;
+	struct list dir;
+	const char *dir_path;
+	char *port_name;
+	int port = -1;
+
+	p = strrchr(path, '/');
+	if (!p) {
+		log_error("Couldn't parse NVMe path to determine port\n");
+		return -1;
+	}
+
+	p++;
+
+	/* p now points to the address, remove the bits after the '.' */
+	f = strchr(p, '.');
+	if (!f) {
+		log_error("Couldn't parse NVMe port address\n");
+		return -1;
+	}
+
+	*f = '\0';
+
+	rc = scan_dir("/sys/bus/pci/slots", &dir);
+	if (rc)
+		return -1;
+
+	list_for_each(&dir, dir_path) {
+		port_name = get_text(dir_path, "address");
+		if (port_name && !strcmp(port_name, p)) {
+			char *dname = strrchr(dir_path, '/');
+
+			dname++;
+			port = atoi(dname);
```

Comment:
Please use strtol instead of atoi. I know it is used in other places in ledmon but it will be fixed soon.